elastic/apm-server
# Enabled licensed fields/features

## Motivation/summary

Enable licensed ("x-pack") fields and beater extensions by:
 - generating a separate `x-pack/apm-server/include/fields.go` file from x-pack fields
 - making the `beat.Creator` (i.e. the function signature which `beater.New` adheres to) parameterisable in the root command

Licensed fields are necessary to avoid errors when running with an unlicensed/OSS-only Elasticsearch. The command changes are necessary to enable us to inject additional features which are dependent on these licensed fields, such as https://github.com/elastic/apm-server/pull/3651.

The fields docs (docs/fields.asciidoc) are built from all fields, like in other Beats.

## Checklist
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

## How to test these changes

```
make apm-server apm-server.x-pack
./apm-server --strict.perms=false export template | grep histogram
./x-pack/apm-server/apm-server --strict.perms=false export template | grep histogram
./x-pack/apm-server/apm-server --strict.perms=false export template --es.version=7.5.0 | grep histogram
```

- OSS build should not have any histogram fields.
- x-pack build should have histogram fields with `--es.version` unspecified
- x-pack build should not have histogram fields with `--es-version=7.5.0` specified. ES 7.5.0 does not support histograms; the [libbeat template generator elides fields with the histogram type for ES <7.6.0](https://github.com/elastic/beats/blob/8e9c73b235119fa517d7d8d2577c18ce35a3aafd/libbeat/template/processor.go#L292-L302).

## Related issues

Relates to https://github.com/elastic/apm-server/pull/3651
